File: platforms/ios/devkit/Sources/NativeblocksDevkit/feature/live/presenter/LiveViewModel.swift
import Combine
import Foundation
import NativeblocksRuntime

internal class LiveViewModel: ObservableObject {
    @Published private(set) var uiState: LiveState = .initial()

    private let authRepository: AuthRepository
    private let liveRepository: LiveRepository
    private let kitEnvironment: DevKitEnvironment
    var cancellables: Set<AnyCancellable> = []

    init(authRepository: AuthRepository, liveRepository: LiveRepository, kitEnvironment: DevKitEnvironment) {
        self.authRepository = authRepository
        self.liveRepository = liveRepository
        self.kitEnvironment = kitEnvironment

        liveRepository.orientedChange()

        authRepository.token()
            .sink { token in
                if !token.isEmpty {
                    if self.uiState.notificationState != .connect && self.uiState.notificationState != .connecting {
                        if kitEnvironment.autoConnect {
                            self.startStream()
                        } else {
                            self.uiState.notificationState = .notConnect
                        }
                    }
                }
            }
            .store(in: &cancellables)

        if authRepository.authCheck() {
            uiState.notificationState = .notConnect
        } else {
            uiState.notificationState = .authRequire
        }

        liveRepository.connectionState()
            .sink { state in
                if authRepository.authCheck() {
                    switch state {
                    case .connected:
                        self.uiState.notificationState = .connect
                    case .connecting:
                        self.uiState.notificationState = .connecting
                    case .notConnected:
                        self.uiState.notificationState = .notConnect
                    }
                } else {
                    self.uiState.notificationState = .authRequire
                }
            }
            .store(in: &cancellables)

        liveRepository.hotReload()
            .sink { route in
                Task {
                    DevKitLogger.debug("LiveViewModel", "HotReload \(route)")
                    if !route.isEmpty {
                        _ = await NativeblocksManager.getInstance(name: kitEnvironment.instanceName).syncFrame(route: route)
                    }
                }
            }
            .store(in: &cancellables)
    }

    private func startStream() {
        if kitEnvironment.screenSharing {
            // NO-OP -> presentScreenSharing()
        }
        liveRepository.start()
    }

    private func endStream() {
        liveRepository.end()
    }

    func updateAction(state: Action) {
        switch state {
        case .connect:
            if authRepository.authCheck() {
                startStream()
            } else {
                uiState.notificationState = .authRequire
            }
        case .disconnect:
            endStream()
        case .start:
            if uiState.notificationState != .connect && uiState.notificationState != .connecting {
                if authRepository.authCheck() {
                    if kitEnvironment.autoConnect {
                        startStream()
                    } else {
                        uiState.notificationState = .notConnect
                    }
                } else {
                    uiState.notificationState = .authRequire
                }
            }
        case .auth:
            AuthScreenSheet.present {
                self.uiState.notificationState = self.uiState.notificationState
            }
        case .logout:
            endStream()
            authRepository.logout()
            AuthScreenSheet.present {
                self.uiState.notificationState = .authRequire
            }
        }
    }
}

File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/feature/live/presenter/LiveViewModel.kt
package io.nativeblocks.devkit.feature.live.presenter

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.nativeblocks.devkit.feature.auth.domain.repository.AuthRepository
import io.nativeblocks.devkit.feature.live.domain.model.ConnectionState
import io.nativeblocks.devkit.feature.live.domain.repository.LiveRepository
import io.nativeblocks.devkit.session.DevKitEnvironment
import io.nativeblocks.runtime.api.NativeblocksManager
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

internal class LiveViewModel(
    private val authRepository: AuthRepository,
    private val liveRepository: LiveRepository,
    private val environment: DevKitEnvironment,
) : ViewModel() {

    private val _uiState = MutableStateFlow(LiveState.initial())
    val uiState: Flow<LiveState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            liveRepository.orientedChange()
        }

        viewModelScope.launch {
            authRepository.token().collect { token ->
                if (token.isNotEmpty()) {
                    if (
                        _uiState.value.notificationState != NotificationState.CONNECT
                        && _uiState.value.notificationState != NotificationState.CONNECTING
                    ) {
                        if (environment.autoConnect) {
                            liveRepository.start()
                        } else {
                            _uiState.update {
                                it.copy(notificationState = NotificationState.NOT_CONNECT)
                            }
                        }
                    }
                }
            }
            if (authRepository.authCheck()) {
                _uiState.update {
                    it.copy(notificationState = NotificationState.NOT_CONNECT)
                }
            }
        }

        viewModelScope.launch {
            liveRepository.connectionState().collect { state ->
                when (state) {
                    ConnectionState.Connected -> {
                        _uiState.update {
                            it.copy(notificationState = NotificationState.CONNECT)
                        }
                    }

                    ConnectionState.Connecting -> {
                        _uiState.update {
                            it.copy(notificationState = NotificationState.CONNECTING)
                        }
                    }

                    ConnectionState.NotConnected -> {
                        _uiState.update {
                            it.copy(notificationState = NotificationState.NOT_CONNECT)
                        }
                    }
                }
            }
        }

        viewModelScope.launch {
            liveRepository.hotReload().collect { frameRoute ->
                NativeblocksManager.getInstance(environment.instanceName).syncFrame(route = frameRoute)
            }
        }
    }

    fun updateAction(state: Action) {
        viewModelScope.launch {
            when (state) {
                Action.CONNECT -> {
                    liveRepository.start()
                }

                Action.DISCONNECT -> {
                    liveRepository.end()
                }

                Action.START -> {
                    if (
                        _uiState.value.notificationState != NotificationState.CONNECT
                        && _uiState.value.notificationState != NotificationState.CONNECTING
                    ) {
                        if (authRepository.authCheck()) {
                            if (environment.autoConnect) {
                                liveRepository.start()
                            } else {
                                _uiState.update {
                                    it.copy(notificationState = NotificationState.NOT_CONNECT)
                                }
                            }
                        } else {
                            _uiState.update {
                                it.copy(notificationState = NotificationState.AUTH_REQUIRE)
                            }
                        }
                    }
                }

                Action.LOGOUT -> {
                    liveRepository.end()
                    authRepository.logout()
                    _uiState.update {
                        it.copy(notificationState = NotificationState.AUTH_REQUIRE)
                    }
                }
            }
        }
    }
}